lib/logstash/filters/yaml.rb

```diff
@@ -58,43 +59,52 @@ def filter(event)
 
     return unless event.include?(@source)
 
-    source = event[@source]
+    source = event.get(@source)
 
-    if @target.nil?
-      # Default is to write to the root of the event.
-      dest = event.to_hash
-    else
-      if @target == @source
-        # Overwrite source
-        event[@target] = {}
-      else
-        event[@target] ||= {}
-      end
-      dest = event[@target]
+    begin
+      parsed = YAML::load(source)
+    rescue => e
+      event.tag("_yamlparsefailure")
+      @logger.warn("Error parsing yaml", :source => @source,
+                   :raw => event.get(@source), :exception => e.message)
+      return
     end
 
-    begin
-      dest.merge!(YAML::load(source))
+    if @target
+      event.set(@target, parsed)
+    else
+      unless parsed.is_a?(Hash)
+        event.tag("_yamlparsefailure")
+        @logger.warn("Parsed YAML object/hash requires a target configuration option", :source => @source, :raw => source)
+        return
+      end
 
-      # If no target, we target the root of the event object. This can allow
-      # you to overwrite @timestamp and this will typically happen for yaml
-      # LogStash Event deserialized here.
-      if !@target && event.timestamp.is_a?(String)
-        event.timestamp = LogStash::Timestamp.parse_iso8601(event.timestamp)
+      # The following logic was copied from Json filter
+      # a) since the parsed hash will be set in the event root, first extract any @timestamp field to properly initialized it
+      parsed_timestamp = parsed.delete(LogStash::Event::TIMESTAMP)
+      begin
+        timestamp = parsed_timestamp ? LogStash::Timestamp.coerce(parsed_timestamp) : nil
+      rescue LogStash::TimestampParserError => e
+        timestamp = nil
       end
 
-      filter_matched(event)
-    rescue => e
-      tag = "_yamlparsefailure"
-      event["tags"] ||= []
-      event["tags"] << tag unless event["tags"].include?(tag)
-      @logger.warn("Trouble parsing yaml", :source => @source,
-                   :raw => event[@source], :exception => e)
-      return
+      # b) then set all parsed fields in the event
+      parsed.each{|k, v| event.set(k, v)}
+
+      # c) finally re-inject proper @timestamp
+      if parsed_timestamp
+        if timestamp
+          event.timestamp = timestamp
+        else
+          event.timestamp = LogStash::Timestamp.new
```

Comment:
wonder if this is actually necessary? I mean, the event already has a timestamp set so if the parsed timestamp is not in a valid format, shouldn't we keep the original one instead of setting in to *now*?

Reply:
I used the same logic present in the json filter here. I think we can discuss still argue the right thing to do, but IMO is out of scope for this PR (whose goal is to make the plugin work again in recent logstashes).